core/src/test/java/smile/projection/ICATest.java

```diff
@@ -0,0 +1,167 @@
+/*******************************************************************************
+ * Copyright (c) 2010 Haifeng Li
+ *   
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *  
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package smile.projection;
+
+import static org.junit.Assert.assertTrue;
+
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import smile.data.AttributeDataset;
+import smile.data.NominalAttribute;
+import smile.data.NumericAttribute;
+import smile.data.parser.ArffParser;
+import smile.data.parser.DelimitedTextParser;
+import smile.math.matrix.DenseMatrix;
+
+/**
+ *
+ * @author rayeaster
+ */
+public class ICATest {
+
+    public ICATest() {
+    }
+
+    @BeforeClass
+    public static void setUpClass() throws Exception {
+    }
+
+    @AfterClass
+    public static void tearDownClass() throws Exception {
+    }
+
+    @Before
+    public void setUp() {
+    }
+
+    @After
+    public void tearDown() {
+    }
+
+    /**
+     * Test of learn method, of FastICA.
+     */
+    @Test
+    public void testWithMixedSignal() {
+        System.out.println("learn ICA with mixed signal...");
+
+        DelimitedTextParser parser = new DelimitedTextParser();
+        parser.setDelimiter(",");
+        parser.setResponseIndex(new NominalAttribute("class"), 3);
+        try {
+            AttributeDataset mixed = parser
+                    .parse(smile.data.parser.IOUtils.getTestDataFile("projection/mixed-signal/test_ica_X.csv"));
+            double[][] x = mixed.toArray(new double[mixed.size()][]);
+
+            parser.setResponseIndex(new NominalAttribute("class"), 1);
+            AttributeDataset signal1 = parser
+                    .parse(smile.data.parser.IOUtils.getTestDataFile("projection/mixed-signal/test_ica_s1.csv"));
+            double[][] s1 = signal1.toArray(new double[signal1.size()][]);// sinusoidal
```

Comment:
Are `s1`, `s2` and `s3` expected signals?

Reply:
@haifengl  yes, the original signal data. I extract the data from the code here: http://scikit-learn.org/stable/auto_examples/decomposition/plot_ica_blind_source_separation.html#sphx-glr-auto-examples-decomposition-plot-ica-blind-source-separation-py